pkg/controllers/execution/execution_controller.go

```diff
@@ -77,31 +94,76 @@ func (c *Controller) Reconcile(ctx context.Context, req controllerruntime.Reques
 		return controllerruntime.Result{Requeue: true}, err
 	}
 
-	if !work.DeletionTimestamp.IsZero() {
-		// Abort deleting workload if cluster is unready when unjoining cluster, otherwise the unjoin process will be failed.
-		if util.IsClusterReady(&cluster.Status) {
-			err := c.tryDeleteWorkload(clusterName, work)
-			if err != nil {
-				klog.Errorf("Failed to delete work %v, namespace is %v, err is %v", work.Name, work.Namespace, err)
-				return controllerruntime.Result{Requeue: true}, err
-			}
-		} else if cluster.DeletionTimestamp.IsZero() { // cluster is unready, but not terminating
-			return controllerruntime.Result{Requeue: true}, fmt.Errorf("cluster(%s) not ready", cluster.Name)
-		}
-
-		return c.removeFinalizer(work)
-	}
-
 	if !util.IsClusterReady(&cluster.Status) {
 		klog.Errorf("Stop sync work(%s/%s) for cluster(%s) as cluster not ready.", work.Namespace, work.Name, cluster.Name)
 		return controllerruntime.Result{Requeue: true}, fmt.Errorf("cluster(%s) not ready", cluster.Name)
 	}
 
-	return c.syncWork(clusterName, work)
+	err = c.ResourceInformer.BuildResourceInformers(cluster, work, c.eventHandler)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	c.worker.Add(req)
+
+	return controllerruntime.Result{}, nil
+}
+
+func (c *Controller) onUpdate(old, cur interface{}) {
+	oldObj := old.(*unstructured.Unstructured)
+	curObj := cur.(*unstructured.Unstructured)
+
+	oldObjCopy := oldObj.DeepCopy()
+	curObjCopy := curObj.DeepCopy()
+
+	clusterName, _ := names.GetClusterName(curObjCopy.GetLabels()[workv1alpha1.WorkNamespaceLabel])
+	if clusterName == "" {
+		return
+	}
+
+	if c.ObjectWatcher.NeedsUpdate(clusterName, oldObjCopy, curObjCopy) {
```

Comment:
NeedsUpdate is used to determine the differences between resources on the control plane and resources in the member cluster. It may not be appropriate to use it directly here. Can we directly compare the spec, annotation, and label of the resource instead?

Reply:
The `ObjectWather.NeedsUpdate` is used in `workstatus controller` to determine if the object in member cluster needs to sync by comparing the `generation` in current object with that last updated recorded in RAM, and comparing the labels, (would be labels and annotations managed by `karmada` after #3804), between the object declared in `work` and that in member cluster. 

In fact, only labels(and annotations later) in desired object is useful for comparation. In a normal workflow, we consider the object in member cluster always the same with our desired. So if we treat the old object in an update event as the desired object, I think it matches the description you gave above.

Reply:
Thanks for explaining. Now that we have the updated resources before and after, can we directly compare their generation? So can we consider deleting the version stored in memory?